earth_enterprise/src/fusion/autoingest/StorageManager.h

```diff
@@ -286,5 +289,61 @@ bool StorageManager<AssetType>::SaveDirtyToDotNew(
   return true;
 }
 
+template<class Asset>
+Asset Find(const std::string & ref, const AssetDefs::Type & type, const std::string & subtype)
+{
+  notify(NFY_WARN, "Find: %s\t%s\t%s", ref.c_str(), ToString(type).c_str(), subtype.c_str());
+  try {
+    Asset asset(ref);
+    notify(NFY_WARN, "Find: %s\t%s", ToString(asset->type).c_str(), asset->subtype.c_str());
+    if (asset &&
+        (asset->type == type) &&
+        (asset->subtype == subtype)) {
+        return Asset(SharedString(ref));
+    }
+  } catch (...) {
+      // do nothing - don't even generate any warnings
+  }
+  return Asset();
+}
+
+template<class Version>
+Version FindVersion(const std::string & ref, const AssetDefs::Type & type, const std::string & subtype)
+{
+    notify(NFY_WARN, "Version: %s\t%s\t%s", ref.c_str(), ToString(type).c_str(), subtype.c_str());
+    try {
+        Version version(ref);
+        notify(NFY_WARN, "Version: %s\t%s", ToString(version->type).c_str(), version->subtype.c_str());
+        if (version &&
+            (version->type == type) &&
+            (version->subtype == subtype)) {
+            return Version(SharedString(ref));
+        }
+    } catch (...) {
+        // do nothing - don't even generate any warnings
+    }
+    return Version();
+}
+
+template<class Asset, class Version>
+void ValidateRefForInput(const std::string & ref, const AssetDefs::Type & type, const std::string & subtype)
```

Comment:
It occurs to me that the asset and version types passed in here have to match, so a programmer that's not being careful enough could write code that will never work, and I don't think the compiler will find the error. Thinking out loud, we could get around that by defining the version type inside the asset class:
```
class CombinedRPAssetD {
  public:
    using Version = CombinedRPAssetVersionD;
};
```
Then you can just have one template parameter for the asset type and get the version parameter from that with `Asset::Version`.

I'll leave it up to you to decide whether to implement that here.